src/altrep-sparse-real.c

```diff
@@ -0,0 +1,176 @@
+}
+
+// -----------------------------------------------------------------------------
+// ALTVEC
+
+void* altrep_sparse_real_Dataptr(SEXP x, Rboolean writeable) {
+  return STDVEC_DATAPTR(alrep_sparse_real_Materialize(x));
+}
+
+const void* altrep_sparse_real_Dataptr_or_null(SEXP vec) {
+    SEXP out = R_altrep_data2(vec);
+
+  if (out == R_NilValue) {
+    return NULL;
+  } else {
+    return out;
+  }
+}
+
+static SEXP altrep_sparse_real_Extract_subset(SEXP x, SEXP indx, SEXP call) {
+
+  SEXP data1 = R_altrep_data1(x);
+  SEXP val_old = VECTOR_ELT(data1, 0);
+  SEXP pos_old = VECTOR_ELT(data1, 1);
+  SEXP matches = Rf_match(pos_old, indx, R_NaInt);
+
+  int n = 0;
+
+  for (int i = 0; i < Rf_length(matches); ++i) { 
+    if (INTEGER_ELT(matches, i) != R_NaInt) {
+      n++;
+    }
+  }
+
+  SEXP val_new = PROTECT(Rf_allocVector(REALSXP, n));
+  SEXP pos_new = PROTECT(Rf_allocVector(INTSXP, n));
+ 
+  int step = 0;
+  int what_pos = 1;
+
+  for (int i = 0; i < Rf_length(matches); ++i) {
+
+    int match = INTEGER_ELT(matches, i);
+    if (match != R_NaInt) {
+      SET_REAL_ELT(val_new, step, REAL_ELT(val_old, match - 1));
+
+      for (int j = 0; j < Rf_length(matches); ++j) {
+        if (INTEGER_ELT(indx, j) == INTEGER_ELT(pos_old, match - 1)) {
+          break;
+        } else {
+          what_pos++;
+        }
+      }
+      SET_INTEGER_ELT(pos_new, step, what_pos);
+      what_pos = 1;
+      step++;
+    }
+  }
+
+  const char *names[] = {"val", "pos", "length", ""};
+  SEXP res = PROTECT(Rf_mkNamed(VECSXP, names)); 
+  SET_VECTOR_ELT(res, 0, val_new);
+  SET_VECTOR_ELT(res, 1, pos_new);
+  SET_VECTOR_ELT(res, 2, Rf_ScalarReal(Rf_length(matches)));
+
+  UNPROTECT(3);
+
+  return ffi_altrep_new_sparse_real(res);
+}
+
+// -----------------------------------------------------------------------------
+// ALTREP
+
+R_xlen_t altrep_sparse_real_Length(SEXP x) {
+  double out = Rf_asReal(VECTOR_ELT(R_altrep_data1(x),2));
+
+  return out;
+}
+
+// What gets printed when .Internal(inspect()) is used
+Rboolean altrep_sparse_real_Inspect(SEXP x,
+                                    int pre,
+                                    int deep,
+                                    int pvec,
+                                    void (*inspect_subtree)(SEXP, int, int, int)) {
+  Rprintf("sparsevctrs_altrep_sparse_real (materialized=%s)\n",
+          R_altrep_data2(x) != R_NilValue ? "T" : "F");
+  return TRUE;
+}
+
+// -----------------------------------------------------------------------------
+// ALTREAL
+
+static double altrep_sparse_real_Elt(SEXP x, R_xlen_t i) {
+
+  if (i > Rf_asReal(VECTOR_ELT(R_altrep_data1(x), 2))) {
+    return NA_REAL;
+  }
+
+  SEXP data1 = R_altrep_data1(x);
+  SEXP val = VECTOR_ELT(data1, 0);
+  SEXP pos = VECTOR_ELT(data1, 1);
+  
+  const R_len_t n = Rf_length(val);
+
+  double out = 0;
+
+  for (int j = 0; j < n; ++j) {
+    if (INTEGER_ELT(pos, j) == i + 1) {
+      out = REAL_ELT(val, j);
+      break;
+    }
+  }
```

Comment:
IIUC the algorithm is:
- Initialize `out` to the default value, `0`
- Loop through `pos` to try and "find" `i + 1` to see if that location has a nonstandard value
- If we find it, exact out the value from `val` and set `out` to it

I think you can go from a linear search to an extremely fast binary search if you require that `pos` be in _sorted_ order up front. That would require some additional up front checks to ensure that `pos` is:
- Sorted
- Has no duplicates

But that would be 100% worth it for this kind of thing to go from `O(n)` to `O(log n)` for element extraction